pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,178 @@
+	var requests []reconcile.Request
+	// Cast the received object into Route object.
+	routeObject := obj.(*routev1.Route)
+
+	// Iterate through the related Ingress Controllers.
+	for _, ic := range routeObject.Status.Ingress {
+		log.Info("queueing ingresscontroller", "name", ic.RouterName)
+		// Create a reconcile.Request for the Ingress Controller.
+		request := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      ic.RouterName,
+				Namespace: r.Namespace,
+			},
+		}
+		requests = append(requests, request)
+	}
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	cache     cache.Cache
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found", "request", request)
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// List all the Namespaces which matches the Namespace LabelSelector.
+	namespaceList := corev1.NamespaceList{}
+	if ingressController.Spec.NamespaceSelector != nil {
+		namespaceLabelSelector := ingressController.Spec.NamespaceSelector.MatchLabels
+		if err := r.cache.List(ctx, &namespaceList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(namespaceLabelSelector),
+		}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list Namespaces %q: %v", request, err)
+		}
+	} else {
+		// If Namespace LabelSelector is not provided then add Namespace with empty name to indicate all Namespaces.
+		namespaceList.Items = append(namespaceList.Items, corev1.Namespace{})
+	}
+
+	// Get the Route LabelSelector from the Ingress Controller.
+	var routeLabelSelector map[string]string
+	if ingressController.Spec.RouteSelector != nil {
+		routeLabelSelector = ingressController.Spec.RouteSelector.MatchLabels
+	}
+
+	// Variable to store the number of routes admitted by the Shard (Ingress Controller).
+	routesAdmitted := 0
+
+	// Iterate through all the Namespaces.
+	for _, namespace := range namespaceList.Items {
+		// List all the Routes which matches the Route LabelSelector and the Namespace.
```

Comment:
A shard can be composed of routes that match the namespace LabelSelector OR match the route LabelSelector.  It does not have to match on both.  E.g. an ingress controller can offer a match on one of x namespaces or one of y route selectors.   See https://docs.openshift.com/container-platform/4.11/networking/ingress-operator.html#nw-ingress-sharding_configuring-ingress

Please add unit tests for this scenario.

Reply:
The docs only provides example of 2 cases where shards are created either for Namespace LabelSelector or Route LabelSelector. However, the actual implementation takes into account both the LabelSelectors while selecting the Routes for the Router.

Reply:
> Please add unit tests for this scenario.

We're using cache to fetch the Route objects inside the reconcile loop. If we were using a client, the fake client package could have been used for unit testing. Can you suggest how we can unit test this part of the code where the cache is being used?

NB: The portion which doesn't need a cache is already extracted out in `routeAdmitted` function for which a unit test is already added.

Reply:
The unit test specification: create an ingress controller with both route and namespace selectors, then create one route with just a route selector and one with just a namespace selector.  Both routes should be counted.

Reply:
> Can you suggest how we can unit test this part of the code where the cache is being used?

The `cache` being used is the problem here for unit test. If we were using `client`, we could've used `fake client` in the unit test. Can you point me to any existing unit test which uses `cache`? I can then add the unit test specification that you have mentioned in the above comment, as the Namespace and Route LabelSelectors are used in the `cache.List` function.

Reply:
> a shard can be composed of routes that match the namespace LabelSelector OR match the route LabelSelector. It does not have to match on both. See https://docs.openshift.com/container-platform/4.11/networking/ingress-operator.html#nw-ingress-sharding_configuring-ingress

An ingresscontroller can specify both a namespace label-selector and a route label-selector, and then the ingresscontroller admits a route only if both the route matches the route label-selector and the route is in a namespace that matches the namespace label-selector.  The documentation might need to be clarified.  

> The `cache` being used is the problem here for unit test. If we were using `client`, we could've used `fake client` in the unit test. Can you point me to any existing unit test which uses `cache`? I can then add the unit test specification that you have mentioned in the above comment, as the Namespace and Route LabelSelectors are used in the `cache.List` function.

I can't find anything in controller-runtime short of using the envtest package.  You could write a fake cache that wrapped a fake client and implemented `List`, but there ought to be a simpler way.  You could try asking the Portfolio Enablement team.

Reply:
> I can't find anything in controller-runtime short of using the envtest package. You could write a fake cache that wrapped a fake client and implemented `List`, but there ought to be a simpler way. You could try asking the Portfolio Enablement team.

Would it be fine if all these scenarios are tested in e2e tests? I have already tried some e2e tests locally for these scenarios. I have explained how I added the e2e tests in this [comment](https://github.com/openshift/cluster-ingress-operator/pull/779#discussion_r974950400).

Reply:
Testing using E2E is fine if it doesn't increase the total runtime too much.  As the E2E tests that you are adding run in parallel with other tests, they should be fine as long as they don't take an extremely long time (say >15 minutes).